lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+       when is_atom(atom) and is_atom(field) do
+    # elixir checks if context is not match
+    s = add_current_env_to_line(s, call_meta, e)
+
+    env = escape_env_entries(meta, s, e)
+
+    case Map.fetch(env, field) do
+      {:ok, value} -> {value, s, e}
+      :error -> {{{:., dot_meta, [escape_map(env), field]}, call_meta, []}, s, e}
+    end
+  end
+
+  # Quote
+
+  defp do_expand({unquote_call, meta, [arg]}, s, e)
+       when unquote_call in [:unquote, :unquote_splicing] do
+    # elixir raises here unquote_outside_quote
+    # we may have cursor there
+    {arg, s, e} = expand(arg, s, e)
+    s = s |> add_current_env_to_line(meta, e)
+    {{unquote_call, meta, [arg]}, s, e}
+  end
+
+  defp do_expand({:quote, meta, [opts]}, s, e) when is_list(opts) do
+    case Keyword.pop(opts, :do) do
+      {nil, _} ->
+        # elixir raises here missing_option
+        # generate a fake do block
+        expand({:quote, meta, [opts, [{:do, {:__block__, [], []}}]]}, s, e)
+
+      {do_block, new_opts} ->
+        expand({:quote, meta, [new_opts, [{:do, do_block}]]}, s, e)
+    end
+  end
+
+  defp do_expand({:quote, meta, [arg]}, s, e) do
+    # elixir raises here invalid_args
+    # we may have cursor there
+    {arg, s, e} = expand(arg, s, e)
+    s = s |> add_current_env_to_line(meta, e)
+    {{:quote, meta, [arg]}, s, e}
+  end
+
+  defp do_expand({:quote, meta, [opts, do_block]}, s, e) when is_list(do_block) do
+    exprs =
+      case Keyword.fetch(do_block, :do) do
+        {:ok, expr} ->
+          expr
+
+        :error ->
+          # elixir raises here missing_option
+          # try to recover from error by generating a fake do block
+          {:__block__, [], [do_block]}
+      end
+
+    valid_opts = [:context, :location, :line, :file, :unquote, :bind_quoted, :generated]
+    {e_opts, st, et} = expand_opts(valid_opts, opts, s, e)
+
+    context = Keyword.get(e_opts, :context, e.module || :"Elixir")
+
+    {file, line} =
+      case Keyword.fetch(e_opts, :location) do
+        {:ok, :keep} -> {e.file, true}
+        :error -> {Keyword.get(e_opts, :file, nil), Keyword.get(e_opts, :line, false)}
+      end
+
+    {binding, default_unquote} =
+      case Keyword.fetch(e_opts, :bind_quoted) do
+        {:ok, bq} ->
+          if is_list(bq) do
+            # safe to drop, opts already expanded
+            bq = Enum.filter(bq, &match?({key, _} when is_atom(key), &1))
+            {bq, false}
+          else
+            {[], false}
+          end
+
+        :error ->
+          {[], true}
+      end
+
+    unquote_opt = Keyword.get(e_opts, :unquote, default_unquote)
+    generated = Keyword.get(e_opts, :generated, false)
+
+    # TODO this is a stub only
+    # res = expand_quote(exprs, st, et)
+    # res |> elem(0) |> IO.inspect
+    # res
+    {q, q_context, q_prelude} =
+      __MODULE__.Quote.build(meta, line, file, context, unquote_opt, generated, et)
```

Comment:
Do you really need to build the quoted expression? Does the compiler care about returned results or do you only care about building the proper environment?

If you really need this functionality, we should find ways to expose it from Elixir. You should not re-implement this part of the compiler. :)

Reply:
I'm not sure yet. In the beginning I did it to verify if my expansion is correct. Now I wonder if at some point we would need true compilation and evaluation